core/deployment/src/main/java/io/quarkus/deployment/configuration/BuildTimeConfigurationReader.java

```diff
@@ -1117,6 +1121,40 @@ public Set<String> getPropertyNames() {
             return properties;
         }
 
+        /**
+         * Use this Config instance to record the runtime default values. We cannot use the main Config
+         * instance because it may record values coming from local development sources (Environment Variables,
+         * System Properties, etc.) in at build time. Local config source values may be completely different between the
+         * build environment and the runtime environment, so it doesn't make sense to record these.
+         *
+         * @return a new SmallRye instance without the EnvSources.
+         */
+        private SmallRyeConfig getConfigForRuntimeRecording() {
+            SmallRyeConfigBuilder builder = ConfigUtils.emptyConfigBuilder();
+            builder.getSources().clear();
+            builder.getSourceProviders().clear();
+            builder.setAddDefaultSources(false)
+                    // Customizers may duplicate sources, but not much we can do about it, we need to run them
+                    .addDiscoveredCustomizers()
+                    // Read microprofile-config.properties, because we disabled the default sources
+                    .withSources(classPathSources(META_INF_MICROPROFILE_CONFIG_PROPERTIES, classLoader));
+
+            // TODO - Should we reset quarkus.config.location to not record from these sources?
```

Comment:
As for `quarkus.config.locations`, it's not that easy to decide what to do.

But it resonates with the comment I made above: I don't fully understand why we need to register defaults for runtime values?
My understanding would be that they should be fully resolved at runtime with whatever config sources are around at runtime.

Reply:
It depends on the phase. We need to record the values for build time - runtime fixed to ensure they do not change between build and runtime.

For runtime, it has been a convenience; to flatten the sources to reduce lookups, avoid loading the config files, and ensure that Quarkus would still run even if a property was missing by falling back to the recorded value from the build.

Reply:
Yeah, so I'm not aware of all the consequences and maybe it could have some consequences but my intuition tells me to drop the recording for runtime and resolve things at runtime.

Of course, I would try to avoid reading the `application.properties` we have in the app and have it preprocessed but I wouldn't go further than that.

Now, I might be missing something entirely and it might cause issues I don't foresee, be it for performance or other things.

Reply:
We can certainly do that. I don't foresee any big consequences, and regarding performance, it is hard to tell the impact without measuring it, but we should be able to figure out something.

We must still record the runtime defaults from roots, mappings, and build steps, but we don't need the config instance in that case.